### PR TITLE
Fixed #32275 -- Added scrypt password hasher.

### PR DESCRIPTION
**Pull request by ryowright**
https://code.djangoproject.com/ticket/32275

### CODE REVIEW

File: 5.1.0 [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Collecting isort
+  Downloading isort-5.6.4-py3-none-any.whl (98 kB)
+Installing collected packages: isort
+Successfully installed isort-5.6.4
```
Comment (jacobtylerwalls) [inline 5.1.0:4]: Hi there. Please check "Has patch" on the ticket to get in the review queue. Also please mind what you check into the repo; this file shouldn't be here. Cheers.

Comment (jacobtylerwalls) [reply 5.1.0:4]: Now the file is back...

Comment (ryowright) [reply 5.1.0:4]: I'm sorry how would I remove this. Do I just delete it on Github?

Comment (jacobtylerwalls) [reply 5.1.0:4]: No problem! That would be one way. Whatever platform you're using to stage your changes should also give you control over which files to avoid staging.

Comment (ryowright) [reply 5.1.0]: Ok I was able to remove the file and push the updated content. I just recently learned to use git reset to unstage it. Thanks for your help.

File: django/utils/crypto.py [hunk 2]
```diff
@@ -87,3 +87,11 @@ def pbkdf2(password, salt, iterations, dklen=0, digest=None):
     password = force_bytes(password)
     salt = force_bytes(salt)
     return hashlib.pbkdf2_hmac(digest().name, password, salt, iterations, dklen)
+
+
+def scrypt(password, salt, n=16384, r=8, p=1, maxmem=0, dklen=64):
```
Comment (apollo13) [inline django/utils/crypto.py:92]: I do not think we need this function.

File: docs/topics/auth/passwords.txt [hunk 3]
```diff
@@ -99,6 +100,7 @@ To use Argon2 as your default storage algorithm, do the following:
             'django.contrib.auth.hashers.PBKDF2PasswordHasher',
             'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
             'django.contrib.auth.hashers.BCryptSHA256PasswordHasher',
+	    'django.contrib.auth.hashers.ScryptPasswordHasher',
```
Comment (apollo13) [inline docs/topics/auth/passwords.txt:103]: Indentation…

File: django/contrib/auth/hashers.py [hunk 4]
```diff
@@ -703,3 +703,60 @@ def safe_summary(self, encoded):
...
+        }
+
+    def must_update(self, encoded):
+        decoded = self.decode(encoded)
+        return decoded['work_factor'] != self.work_factor
+
+    def harden_runtime(self, password, encoded):
+        decoded = self.decode(encoded)
+        extra_iterations = self.work_factor - decoded['work_factor']
+        if extra_iterations > 0:
```
Comment (apollo13) [inline django/contrib/auth/hashers.py:760]: What about `r` and `p` here?

**Review by apollo13 — CHANGES_REQUESTED** [review]

File: django/conf/global_settings.py [hunk 5]
```diff
@@ -529,6 +529,7 @@ def gettext_noop(s):
     'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
     'django.contrib.auth.hashers.Argon2PasswordHasher',
     'django.contrib.auth.hashers.BCryptSHA256PasswordHasher',
+    'django.contrib.auth.hashers.ScryptPasswordHasher',
```
Comment (apollo13) [inline django/conf/global_settings.py:532]: Does it make sense to add the hasher here? It will not get used anyways *scratches head*

File: django/contrib/auth/hashers.py (hunk 4 above)
Comment (ryowright) [reply django/contrib/auth/hashers.py:760]: I thought the harden_runtime function was based only off of the work_factor, according to lines 235-236 in the hashers.py file.

Comment (apollo13) [reply django/contrib/auth/hashers.py:760]: I "depends" :) `harden_runtime` is there to bridge the CPU timing difference between the old and new settings, "work factor" in that comment might have been the wrong wording due to the fact that hasher like `scrypt` actually have a parameter called similar. So the main question here is whether `r` and `p` contribute to the runtime in a way that the should be included here (and how, ie linear or quadratic). According to https://blog.filippo.io/the-scrypt-parameters/ it says the following for `r`: 
> It does that by iterating the hash function 2𝑟 times, so both memory usage (to store the hash values) and CPU time scale linearly with it. That is, if 𝑟 doubles the resources double.

For `p` it says:
> 𝑝 > 1 can be handled in two ways: sequentially, which does not increase memory usage but requires 𝑝 times the CPU and wall clock time; or parallelly, which requires 𝑝 times the memory and effective CPU time, but does not increase wall clock time.

So it will be hard to include `p` in `harden_runtime` (I guess we can ignore it and pretend it does not change the wall time because the server has enough CPU -- probably not true but including the number of available CPU threads in this calculation seems overkill).

Comment (ryowright) [reply django/contrib/auth/hashers.py:760]: Okay I see. I added the block_size to the runtime but left the parallelism out. Thanks for your help.

File: django/utils/crypto.py (hunk 2 above)
Comment (ryowright) [reply django/utils/crypto.py:92]: The `scrypt ` function converts the string password and salt into bytes. Should I use the `force_bytes` method in the hashers.py scrypt class then? I just followed the format of the pbkdf2 function above the `scrypt` function I defined.

File: django/contrib/auth/hashers.py [hunk 6]
```diff
@@ -703,3 +703,60 @@ def safe_summary(self, encoded):
...
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = "scrypt"
+    work_factor = 16384
+    block_size = 8
+    parallelism = 1
+
+    def encode(self, password, salt, n=16384, r=8, p=1):
```
[the remark this answers is not in the document]
Comment (ryowright) [reply django/contrib/auth/hashers.py:719]: The `harden_runtime` method uses `extra_iterations` and `extra_block` as possible parameter values to the `encode ` method so I was wondering if I would still need to remove the `n` and `r`. I did remove the `p` parameter.

File: django/conf/global_settings.py (hunk 5 above)
Comment (ryowright) [reply django/conf/global_settings.py:532]: I removed the line from the global_settings.py.

File: docs/topics/auth/passwords.txt (hunk 3 above)
Comment (ryowright) [reply docs/topics/auth/passwords.txt:103]: Added indentation.

File: django/contrib/auth/hashers.py [hunk 7]
```diff
@@ -703,3 +703,60 @@ def safe_summary(self, encoded):
...
+    def safe_summary(self, encoded):
+        decoded = self.decode(encoded)
+        return {
+            _('algorithm'): decoded['algorithm'],
+            _('work_factor'): decoded['work_factor'],
+            _('salt'): mask_hash(decoded['salt']),
+            _('hash'): mask_hash(decoded['hash']),
+        }
+
+    def must_update(self, encoded):
```
[the remark this answers is not in the document]
Comment (ryowright) [reply django/contrib/auth/hashers.py:753]: Added ability to update based on `block_size` and `parallelism`.

File: django/contrib/auth/hashers.py [hunk 8]
```diff
@@ -703,3 +703,60 @@ def safe_summary(self, encoded):
...
+            'parallelism': int(parallelism)
+        }
+
+    def verify(self, password, encoded):
+        decoded = self.decode(encoded)
+        encoded_2 = self.encode(password, decoded['salt'], decoded['work_factor'],
+                                decoded['block_size'], decoded['parallelism'])
+        return constant_time_compare(encoded, encoded_2)
+
+    def safe_summary(self, encoded):
```
[the remark this answers is not in the document]
Comment (ryowright) [reply django/contrib/auth/hashers.py:744]: Added `block_size` and `parallelism` to the `safe_summary` method.

**Comment by timgraham** [comment]
Anthony, it sounds like you've made changes some locally but haven't pushed them to your branch.

File: docs/topics/auth/passwords.txt [hunk 9]
```diff
@@ -211,6 +215,34 @@ follows:
...
+
+The default values for the respective attributes are:
+
+#. ``work_factor`` has a default value of 16384.
+#. ``block_size`` has a default value of 8.
+#. ``parallelism`` has a default value of 1.
+
+.. note::
+
+    Scrypt is only available in the Python standard library for Python 3.6 and up.
```
Comment (timgraham) [inline docs/topics/auth/passwords.txt:244]: This isn't necessary as Django doesn't support versions older than 3.6. I see that Python's documentation for `hashlib.script` notes a requirement of OpenSSL 1.1+. It would be worth looking at common systems today to see if that's something that might be worth noting.

**Comment by ryowright** [comment]
Just pushed recent changes @timgraham. I apologize I was distracted by some other stuff I was working on.

File: docs/topics/auth/passwords.txt [hunk 10]
```diff
@@ -62,10 +62,11 @@ The default for :setting:`PASSWORD_HASHERS` is::
         'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
         'django.contrib.auth.hashers.Argon2PasswordHasher',
         'django.contrib.auth.hashers.BCryptSHA256PasswordHasher',
+	'django.contrib.auth.hashers.ScryptPasswordHasher',
```
Comment (timgraham) [inline docs/topics/auth/passwords.txt:65]: This isn't accurate based on your latest edit (removal from `global_settings.py`).

File: django/contrib/auth/hashers.py [hunk 11]
```diff
@@ -704,3 +704,70 @@ def safe_summary(self, encoded):
...
+    algorithm = "scrypt"
+    work_factor = 16384
+    block_size = 8
+    parallelism = 1
+
+    def encode(self, password, salt, n=work_factor, r=block_size):
+        assert password is not None
+        assert salt and '$' not in salt
+        hash = scrypt(password, salt, n=n,
+                      r=r, p=self.parallelism)
```
Comment (timgraham) [inline django/contrib/auth/hashers.py:724]: Throught your patch, please use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/auth_tests/test_hashers.py [hunk 12]
```diff
@@ -596,3 +596,22 @@ def setter(password):
             self.assertTrue(state['upgraded'])
         finally:
             setattr(hasher, attr, old_value)
+
+
+@override_settings(PASSWORD_HASHERS=PASSWORD_HASHERS)
```
Comment (timgraham) [inline tests/auth_tests/test_hashers.py:601]: This won't work anymore without `ScryptPasswordHasher` in `PASSWORD_HASHERS`.

File: django/contrib/auth/hashers.py [hunk 13]
```diff
@@ -704,3 +704,70 @@ def safe_summary(self, encoded):
 
     def harden_runtime(self, password, encoded):
         pass
+
+
+class ScryptPasswordHasher(BasePasswordHasher):
```
Comment (timgraham) [inline django/contrib/auth/hashers.py:709]: I wouldn't sort this at the bottom of the file below all the insecure hashers.

File: tests/auth_tests/test_hashers.py [hunk 14]
```diff
@@ -596,3 +596,22 @@ def setter(password):
...
+class TestUtilsHashPassScrypt(SimpleTestCase):
+
+    def test_scrypt(self):
+        encoded = make_password('lètmein', 'seasalt', 'scrypt')
+        self.assertEqual(
```
Comment (timgraham) [inline tests/auth_tests/test_hashers.py:606]: You can use something like this (wrapping lines at 79 chars):
```
self.assertEqual(
    encoded, 
    'scrypt$16384$seasalt$Qj3+9PPyRjSJIebHnG81TMjsqtaIGxNQG/aEB/NYaf'
    'TJ7tibgfYz71m0ldQESkXFRkdVCBhhY8mx7rQwite/Pw==$8$1'
)
```

**Review by timgraham — COMMENTED** [review]
Tests seem rather lacking, for example, there's no testing of password upgrades.

See also our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) about documentation requirements (release notes and versionadded annotation).

File: django/contrib/auth/hashers.py [hunk 15]
```diff
@@ -704,3 +704,70 @@ def safe_summary(self, encoded):
...
+    """
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = "scrypt"
+    work_factor = 16384
+    block_size = 8
+    parallelism = 1
```
Comment (illia-v) [inline django/contrib/auth/hashers.py:718]: Increasing this parameters slightly may lead to exceeding the default 32-MiB memory limit.

On my machine setting `work_factor` to 32768 (the next valid value after 16384) results in `ValueError: Invalid parameter combination for n, r, p, maxmem.`.

Could you please make it possible to increase the memory limit?

Comment (ryowright) [reply django/contrib/auth/hashers.py]: I added maxmem class attribute to allow for memory limit increase in the most recent commit.

File: django/contrib/auth/hashers.py (hunk 13 above)
Comment (ryowright) [reply django/contrib/auth/hashers.py]: I moved it up to the section with the secure hashers, just under the Bcrypt password hasher.

File: tests/auth_tests/test_hashers.py (hunk 14 above)
Comment (ryowright) [reply tests/auth_tests/test_hashers.py:641]: I adjusted the string to match that in the most recent commit.

File: tests/auth_tests/test_hashers.py (hunk 12 above)
Comment (ryowright) [reply tests/auth_tests/test_hashers.py]: Removed in most recent commit.

File: django/contrib/auth/hashers.py (hunk 11 above)
Comment (ryowright) [reply django/contrib/auth/hashers.py]: Modified indentations to conform to the coding style.

File: docs/topics/auth/passwords.txt (hunk 10 above)
Comment (ryowright) [reply docs/topics/auth/passwords.txt]: Removed `ScryptPasswordHasher ` lines from passwords.txt documentation.

File: docs/topics/auth/passwords.txt (hunk 9 above)
Comment (ryowright) [reply docs/topics/auth/passwords.txt]: Removed Python 3.6 note and added note regarding OpenSSL 1.1+.

File: django/contrib/auth/hashers.py (hunk 15 above)
Comment (illia-v) [reply django/contrib/auth/hashers.py]: Thanks!

I see that you included `maxmem` in a string returned by `encode`. Is there any reason it should be there?
`maxmem` should not affect a hash

Comment (kerkeslager) [reply django/contrib/auth/hashers.py]: ilia-v,

The work factor limitation you've run into is a limitation of the underlying Python standard library.

```~$ python3
Python 3.9.1 (default, Feb  3 2021, 07:04:15) 
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from hashlib import scrypt
>>> scrypt(b'password', salt=b'salt', n=16384, r=8, p=1)
b'tW1\xafD\x84\xf3#\x96\x89i\xed\xa2\x89\xae\xee\x00[Y\x03\xacV\x1ed\xa5\xac\xa1!y{\xf7sN\xf9\xfdXB.."\x18;\xca\xcb\xa9\xec\x87\xba\x0c\x83\xb7\xa2\xe7\x88\xf0<\xe0\xda\x06F43\xcd\xa6'
>>> scrypt(b'password', salt=b'salt', n=32768, r=8, p=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Invalid parameter combination for n, r, p, maxmem.```

I agree that it's a problem, but I think fixing it is outside the scope of Django's responsibility. If you want it fixed take it up with the Python project.

Comment (illia-v) [reply django/contrib/auth/hashers.py]: @kerkeslager `hashlib.scrypt` is a low-level function that takes multiple arguments.

`maxmem` is one of them, it has a default value that does not suit all cases. Therefore, Django should allow changing the value. @ryowright has already made the necessary change.

Note, `maxmem` does not affect generated hashes, its value should not be included in a string saved to a database and should not be compared in `must_update`.

File: django/conf/global_settings.py (hunk 5 above)
Comment (felixxm) [reply django/conf/global_settings.py]: I think it's worth, re-added.

File: django/contrib/auth/hashers.py [hunk 16]
```diff
@@ -512,6 +513,84 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
...
+    maxmem = 0
+
+    def encode(self, password, salt, n=work_factor, r=block_size, maxmem=maxmem):
```
Comment (felixxm) [inline django/contrib/auth/hashers.py:528]: We can remove the `maxmem` argument and use `self.maxmem`.

File: django/contrib/auth/hashers.py (hunk 15 above)
Comment (felixxm) [reply django/contrib/auth/hashers.py]: @illia-v Agreed, I removed `maxmem` from an encoded password.

File: django/contrib/auth/hashers.py [hunk 17]
```diff
@@ -512,6 +513,84 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
...
+    def must_update(self, encoded):
+        decoded = self.decode(encoded)
+        return (
+            decoded['work_factor'] != self.work_factor or
+            decoded['block_size'] != self.block_size or
+            decoded['parallelism'] != self.parallelism or
+            decoded['maxmem'] != self.maxmem
+        )
+
+    def harden_runtime(self, password, encoded):
```
Comment (apollo13) [inline django/contrib/auth/hashers.py:582]: My main worry here is: Is this correct and does it make sense to implement for such a complex hasher (notably we already have others where we argue it is simply not possible in a sensible way).

Since scrypt can raise errors like this:
> ValueError: Invalid parameter combination for n, r, p, maxmem.

I am wondering if `must_update` couldn't also trigger this condition. Or can we always calculate `extra_iterations` and `extra_block` and be sure that the combinations are valid?

File: django/contrib/auth/hashers.py [hunk 18]
```diff
@@ -512,6 +513,84 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
...
 
 
+class ScryptPasswordHasher(BasePasswordHasher):
+    """
+    Secure password hashing using the Scrypt algorithm
+
+    The result is a 64 byte binary string.
+    """
+    algorithm = "scrypt"
+    work_factor = 16384
```
Comment (apollo13) [inline django/contrib/auth/hashers.py:523]: Do we have any plan on how & when to upgrade those as time passes by? (ie like we have for pbkdf2)

File: docs/topics/auth/passwords.txt [hunk 19]
```diff
@@ -98,7 +98,7 @@ To use Argon2 as your default storage algorithm, do the following:
             'django.contrib.auth.hashers.Argon2PasswordHasher',
             'django.contrib.auth.hashers.PBKDF2PasswordHasher',
             'django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher',
-            'django.contrib.auth.hashers.BCryptSHA256PasswordHasher',
+            'django.contrib.auth.hashers.BCryptSHA256PasswordHasher'
```
Comment (apollo13) [inline docs/topics/auth/passwords.txt:101]: Unrelated changes, should be reverted.

File: docs/topics/auth/passwords.txt [hunk 20]
```diff
@@ -365,6 +395,7 @@ The full list of hashers included in Django is::
         'django.contrib.auth.hashers.UnsaltedSHA1PasswordHasher',
         'django.contrib.auth.hashers.UnsaltedMD5PasswordHasher',
         'django.contrib.auth.hashers.CryptPasswordHasher',
+	    'django.contrib.auth.hashers.ScryptPasswordHasher',
```
Comment (apollo13) [inline docs/topics/auth/passwords.txt:398]: indention?

File: django/utils/crypto.py (hunk 2 above)
Comment (felixxm) [reply django/utils/crypto.py]: I agree with Florian, it's enough to call `encode()`.

File: docs/topics/auth/passwords.txt (hunk 19 above)
Comment (felixxm) [reply docs/topics/auth/passwords.txt:101]: I will push my basic edits in 10 minutes.

File: django/contrib/auth/hashers.py [hunk 21]
```diff
@@ -512,6 +512,88 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
...
+            _('salt'): mask_hash(decoded['salt']),
+            _('hash'): mask_hash(decoded['hash']),
+        }
+
+    def must_update(self, encoded):
+        decoded = self.decode(encoded)
+        return (
+            decoded['work_factor'] != self.work_factor or
+            decoded['block_size'] != self.block_size or
+            decoded['parallelism'] != self.parallelism
```
Comment (felixxm) [inline django/contrib/auth/hashers.py:582]: `self._test_scrypt_upgrade('parallelism', 'parallelism', 2)` fails, it seems we shouldn't take `parallelism` into account.

**Review by felixxm — COMMENTED** [review]
@ryowright Thanks for all your efforts :star: Do you have time to keep working on this? If not, I can push this to the finish line.

I rebased, squashed commit, and pushed edits. 

I think we should add a separate section `Using Scrypt with Django` to docs. Also, releases notes are missing and `test_scrypt_upgrade()` doesn't work.

File: django/contrib/auth/hashers.py (hunk 21 above)
Comment (felixxm) [reply django/contrib/auth/hashers.py:582]: We confirmed that `parallelism` should be taken into account.

File: django/contrib/auth/hashers.py [hunk 22]
```diff
@@ -512,6 +512,88 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
...
+            maxmem=self.maxmem,
+            dklen=64,
+        )
+        hash_ = base64.b64encode(hash_).decode('ascii').strip()
+        return '%s$%d$%s$%s$%d$%d' % (
+            self.algorithm, n, salt, hash_, r, self.parallelism,
+        )
+
+    def decode(self, encoded):
+        algorithm, work_factor, salt, hash_, block_size, parallelism = encoded.split('$', 6)
```
Comment (apollo13) [inline django/contrib/auth/hashers.py:547]: Is there a good reason to order the data like this? I'd personally expect the hash to be at the end, so it could include a `$` .

Comment (felixxm) [reply django/contrib/auth/hashers.py:547]: Good point, updated :+1:

File: django/contrib/auth/hashers.py (hunk 18 above)
Comment (felixxm) [reply django/contrib/auth/hashers.py]: I'm not sure, I don't think there is a need to do this per release. :thinking: 

`work_factor` must be a power of two and the OpenSSL limits memory usage to 32 MiB, so we would have to increase `maxmem` as well: :teacher: 
```
(2 ** 15) * (2 * 8) * 64 = 33554432 = 32 MiB
```

File: django/contrib/auth/hashers.py (hunk 17 above)
Comment (felixxm) [reply django/contrib/auth/hashers.py:585]: > My main worry here is: Is this correct and does it make sense to implement for such a complex hasher (notably we already have others where we argue it is simply not possible in a sensible way).
> 
> Since scrypt can raise errors like this:
> 
> > ValueError: Invalid parameter combination for n, r, p, maxmem.
> 
> I am wondering if `must_update` couldn't also trigger this condition. Or can we always calculate `extra_iterations` and `extra_block` and be sure that the combinations are valid?

You're right parameters may no be valid, e.g.
```
self.work_factor = 2 **14
decoded['work_factor'] = 2 ** 11
```
both are a power of 2, however `extra_iterations = 14336` is not and raises `ValueError: n must be a power of 2`.

**Comment by felixxm** [comment]
@apollo13 I pushed final edits with docs changes and added release notes. Can you take a look?

File: docs/releases/4.0.txt [hunk 23]
```diff
@@ -58,6 +58,12 @@ For example::
 Functional unique constraints are added to models using the
 :attr:`Meta.constraints <django.db.models.Options.constraints>` option.
 
+``scrypt`` password hasher
+--------------------------
+
+The new :ref:`scrypt password hasher <scrypt-usage>` is substantially more
```
Comment (apollo13) [inline docs/releases/4.0.txt:64]: If we are recommending it like this (do we really want to?) we should also switch the default (at least in the project template) to be scrpt.

Comment (felixxm) [reply docs/releases/4.0.txt:64]: > If we are recommending it like this (do we really want to?) 

I based it on @alex's opinion in the ticket description. Maybe too enthusiastic :thinking:  

> we should also switch the default (at least in the project template) to be scrpt.

I was thinking about it, but it requires OpenSSL 1.1+ so it may be too early to the change the default for new projects.

Comment (apollo13) [reply docs/releases/4.0.txt:64]: Then I'd add least add a note why it is not the default yet (maybe also tone down the wording a bit). Also the memory requirements can cause problems for some…

Comment (felixxm) [reply docs/releases/4.0.txt]: I adjusted release notes, any suggestions are welcome.

File: docs/releases/4.0.txt [hunk 24]
```diff
@@ -58,6 +58,13 @@ For example::
 Functional unique constraints are added to models using the
 :attr:`Meta.constraints <django.db.models.Options.constraints>` option.
 
+``scrypt`` password hasher
+--------------------------
+
+The new :ref:`scrypt password hasher <scrypt-usage>` is more secure and
+recommended over PBKDF2. However, it's not the default as it requires the
+OpenSSL 1.1+ and much memory which can cause issues.
```
Comment (smithdc1) [inline docs/releases/4.0.txt:66]: 'requires the OpenSSl' -> 'requires OpenSSl'

'much memory' -> 'much more memory' or maybe just 'more memory'.

File: docs/topics/auth/passwords.txt [hunk 25]
```diff
@@ -137,6 +140,41 @@ To use Bcrypt as your default storage algorithm, do the following:
...
+require very much memory, attackers with sufficient resources can launch
+large-scale parallel attacks in order to speed up the attacking process.
+scrypt_ is specifically designed to use large amounts of memory compared to
+other password-based key derivation functions in order to limit the amount of
+parallelism an attacker can use, see :rfc:`7914` for more details.
+
+To use scrypt_ as your default storage algorithm, do the following:
+
+#. Modify :setting:`PASSWORD_HASHERS` to list ``ScryptPasswordHasher`` first.
+   That is, in your settings file, you'd put::
```
Comment (smithdc1) [inline docs/topics/auth/passwords.txt:161]: Not sure about "you'd". Maybe just end the sentence at 'file'.

File: docs/releases/4.0.txt (hunk 24 above)
Comment (jacobtylerwalls) [reply docs/releases/4.0.txt:66]: and a comma before "which"

File: django/contrib/auth/hashers.py [hunk 26]
```diff
@@ -512,6 +512,82 @@ class BCryptPasswordHasher(BCryptSHA256PasswordHasher):
...
+    """
+    algorithm = 'scrypt'
+    block_size = 8
+    maxmem = 0
+    parallelism = 1
+    work_factor = 2 ** 14
+
+    def encode(self, password, salt, n=None, r=None, p=None):
+        assert password is not None
+        assert salt and '$' not in salt
```
Comment (cjerdonek) [inline django/contrib/auth/hashers.py:527]: Is it possible for these asserts (and below) to fail without a bug in Django? If so, proper exceptions should be raised a la: https://code.djangoproject.com/ticket/32508

File: tests/auth_tests/test_hashers.py [hunk 27]
```diff
@@ -632,3 +632,68 @@ def setter(password):
...
+        self.assertIs(check_password(' ', blank_encoded), False)
+
+    def test_scrypt_decode(self):
+        encoded = make_password('lètmein', 'seasalt', 'scrypt')
+        hasher = get_hasher('scrypt')
+        decoded = hasher.decode(encoded)
+        self.assertEqual(decoded['block_size'], hasher.block_size)
+        self.assertEqual(decoded['parallelism'], hasher.parallelism)
+        self.assertEqual(decoded['salt'], 'seasalt')
+        self.assertEqual(decoded['work_factor'], hasher.work_factor)
```
Comment (cjerdonek) [inline tests/auth_tests/test_hashers.py:665]: Subtests can be used here (with pairs `(key, expected)`).

File: tests/auth_tests/test_hashers.py [hunk 28]
```diff
@@ -632,3 +632,68 @@ def setter(password):
...
+            # Password is updated.
+            self.assertIs(check_password('lètmein', encoded, setter, 'scrypt'), True)
+            self.assertIs(state['upgraded'], True)
+        finally:
+            setattr(hasher, attr, old_value)
+
+    def test_scrypt_upgrade(self):
+        self._test_scrypt_upgrade('work_factor', 'work factor', 2 ** 11)
+        self._test_scrypt_upgrade('block_size', 'block size', 10)
+        self._test_scrypt_upgrade('parallelism', 'parallelism', 2)
```
Comment (cjerdonek) [inline tests/auth_tests/test_hashers.py:699]: Subtests can also be used here.

**Review by cjerdonek — COMMENTED** [review]
Some comments.

File: docs/releases/4.0.txt (hunk 24 above)
Comment (smithdc1) [reply docs/releases/4.0.txt:66]: > Or maybe just finish at memory. It's already implied from the however there are "issues".

Yes. 

Or maybe just finish at memory. I think it's implied from the however there are 'issues' with what follows.

File: django/contrib/auth/hashers.py (hunk 26 above)
Comment (felixxm) [reply django/contrib/auth/hashers.py:527]: We decided to leave similar asserts in other hashers. Users should not call `encode()` directly.

Comment (cjerdonek) [reply django/contrib/auth/hashers.py:527]: > Users should not call encode() directly.

I looked, and unlike I'm mistaken, Django's documentation does provide an example that instructs users to create a data migration in which the value in user code gets passed to `encode()` as is: https://docs.djangoproject.com/en/3.2/topics/auth/passwords/#password-upgrading-without-requiring-a-login

**Comment by felixxm** [comment]
Thanks for reviews. I pushed small edits.

File: django/contrib/auth/hashers.py (hunk 26 above)
Comment (felixxm) [reply django/contrib/auth/hashers.py:527]: Yes, but these assertions can fail only for users with an incorrect hash in the database. IMO `assert`s can stay here, if you think otherwise we can discuss changing them in all hashers in ticket-32508.

Comment (cjerdonek) [reply django/contrib/auth/hashers.py:527]: For future reference, I added a comment to the ticket here: https://code.djangoproject.com/ticket/32508#comment:35

Comment (pauloxnet) [reply django/contrib/auth/hashers.py:527]: I also commented the same issue.

Comment (felixxm) [reply django/contrib/auth/hashers.py:527]: OK, I prepared PR #14680 with a common hook that can be re-used here.

Comment (felixxm) [reply django/contrib/auth/hashers.py:532]: Rebased.

**Review by apollo13 — APPROVED** [review]

**Comment by felixxm** [comment]
Thanks y'all for reviews :rocket:
